resources/config/services.yml

```diff
@@ -8,11 +8,11 @@ services:
         class: GrumPHP\Configuration\GrumPHP
         arguments:
           - '@service_container'
+        public: true
 
     event_dispatcher:
-        class: Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher
-        arguments:
-          - '@service_container'
+        class: Symfony\Component\EventDispatcher\EventDispatcher
```

Comment:
Not sure why the ContainerAwareEventDispatcher was used instead of the regular one. Will this break anything?

Reply:
Been lookin' into this one. Looks like the Travis array-related message is coming from over here... :(

This change was related to the deprecation of `ContainerAwareEventDispatcher`, as it's gone from v4. But the thing is, it is still needed for the `RegisterListenersPass` (which is doing some magic, and eventually _can_ call the container-aware's additional `addSubscriberService` method).  
_P.S.: It's `subscriber.stash_unstaged_changes`, btw..._

I was thinking about the fix below, which falls back on the container-aware implementation (v2.8, v3.4) as long as it's available, but uses the newer version when it does not anymore.

It fixes the issue **without having to duplicate Symfony core code**. But however, I'm not able to commit this because of Object Calisthenics violations:
- phpmd's "no else"
- phpparser's `never_use_else`

You'd think an _early return_ alternative (without else) would work then. Doesn't like that either, as it starts complaining about side effects.  
I can silence phpmd (hence the `@codingStandardsIgnoreFile`), but can't find a way to do the same to phpparser.

---
```diff
diff --git a/resources/config/services.yml b/resources/config/services.yml
index ab6eab9..43184d0 100644
--- a/resources/config/services.yml
+++ b/resources/config/services.yml
@@ -11,7 +11,9 @@ services:
         public: true
 
     event_dispatcher:
-        class: Symfony\Component\EventDispatcher\EventDispatcher
+        class: GrumPHP\Event\EventDispatcher
+        arguments:
+          - '@service_container'
         public: true
 
     filesystem:
diff --git a/src/Event/EventDispatcher.php b/src/Event/EventDispatcher.php
new file mode 100644
index 0000000..ccba69c
--- /dev/null
+++ b/src/Event/EventDispatcher.php
@@ -0,0 +1,18 @@
+<?php
+
+namespace GrumPHP\Event;
+
+use Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher;
+use Symfony\Component\EventDispatcher\EventDispatcher as SymfonyEventDispatcher;
+
+// Backward compatible layer for Symfony < 4.0.
+// The container aware variant has been deprecated (3.4) and removed (4.0),
+// but is still required to keep the RegisterListenersPass happy.
+// @codingStandardsIgnoreFile
+if (class_exists(ContainerAwareEventDispatcher::class)) {
+    class EventDispatcher extends ContainerAwareEventDispatcher {
+    }
+} else {
+    class EventDispatcher extends SymfonyEventDispatcher {
+    }
+}
```

Reply:
Another alternative could be:
- restore the original service definition (`ContainerAwareEventDispatcher`)
- using a compiler pass, swap it out with the regular one (plain `EventDispatcher`) if class no longer exists.

Or, the other way around:
- use v4's `EventDispatcher` by default
- swap it out with `ContainerAwareEventDispatcher` when available

---
But, as you didn't like my previous attempt to a compiler pass (visibility), I went with the above extension instead.

Which direction do you suggest, @veewee? Or do you have other alternatives?  
I'd like to avoid duplicating the whole code (and tests) duplication for a  `ContainerAwareEventDispatcher` copy (which is deprecated). Or a `RegisterListenersPass` copy... 👎

Reply:
What about adding a factory class that returns the valid eventdispatcher for the installed symfony version? More info at https://symfony.com/doc/current/service_container/factories.html

Reply:
Ah, ofcourse. That should work too... 👍  
Will look into that, instead. Thanks.

_The break from work-holiday obviously doesn't do me much good – for coding xD_